modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java

```diff
@@ -315,46 +313,35 @@ public void onFailure(Exception e) {
     }
 
     @Override
-    public void deleteSnapshots(
-        Collection<SnapshotId> snapshotIds,
-        long repositoryDataGeneration,
-        IndexVersion repositoryFormatIndexVersion,
-        SnapshotDeleteListener listener
-    ) {
-        final SnapshotDeleteListener wrappedListener;
-        if (SnapshotsService.useShardGenerations(repositoryFormatIndexVersion)) {
-            wrappedListener = listener;
-        } else {
-            wrappedListener = new SnapshotDeleteListener() {
-                @Override
-                public void onDone() {
-                    listener.onDone();
-                }
-
-                @Override
-                public void onRepositoryDataWritten(RepositoryData repositoryData) {
-                    logCooldownInfo();
-                    final Scheduler.Cancellable existing = finalizationFuture.getAndSet(threadPool.schedule(() -> {
-                        final Scheduler.Cancellable cancellable = finalizationFuture.getAndSet(null);
-                        assert cancellable != null;
-                        listener.onRepositoryDataWritten(repositoryData);
-                    }, coolDown, snapshotExecutor));
-                    assert existing == null : "Already have an ongoing finalization " + finalizationFuture;
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    logCooldownInfo();
-                    final Scheduler.Cancellable existing = finalizationFuture.getAndSet(threadPool.schedule(() -> {
-                        final Scheduler.Cancellable cancellable = finalizationFuture.getAndSet(null);
-                        assert cancellable != null;
-                        listener.onFailure(e);
-                    }, coolDown, snapshotExecutor));
-                    assert existing == null : "Already have an ongoing finalization " + finalizationFuture;
-                }
-            };
-        }
-        super.deleteSnapshots(snapshotIds, repositoryDataGeneration, repositoryFormatIndexVersion, wrappedListener);
+    protected SnapshotDeleteListener wrapWithWeakConsistencyProtection(SnapshotDeleteListener snapshotDeleteListener) {
```

Comment:
`S3Repository` made this slightly more involved than expected - previously we were wrapping up the listener with this cooldown thing before starting the delete, but if we want to wait to compute the repo format version then we must do the wrapping later too.

Fortunately this is covered by existing tests, see `org.elasticsearch.repositories.s3.S3BlobStoreRepositoryTests#testEnforcedCooldownPeriod`.

Reply:
Nit: can we just call the original listener as `listener` instead of `snapshotDeleteListener`. It helps both reduce the changeset a little bit and easier to spot which one is the original listener.